{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem:\n",
    "Once we sequenced a cancer tumor, we can identify thousands of mutations. The challenge is to distinguish the mutations that contribute to turmor growth, we call drivers, from the neutral mutations that we can passengers. Provided with an expert-annotated knowledge base, can you develop a ML algorithm that automatically classifies genetic variants?\n",
    "The knowledge base for each annotated variant is a collection of relevant journal articles, done by experts. The assumption is that the text contains enough information for classification. \n",
    "\n",
    "Traingin Data:\n",
    "    The name of the genes, and the corresponding variant type are categorized into 9 classes, based on a large body of journal texts. The meanings of the 9 classes are unknown during the competition. I listed below because it helps to understand the classification.\n",
    "    External knowledge about the 9 classes:\n",
    "    7:gain of function, 2:likely gof, 3:neutral, 5:likely neutral, 6:inconclusive, 1:likely loss of func., 4:lof, 8:likely swithc off func, 9: sof\n",
    "\n",
    "Use pandas, numpy, xgboost, sklearn, matplotlib.\n",
    "\n",
    "Hightlights:\n",
    "* Feature analysis:  \n",
    "  Use regular expression for text search and counting occurences of gene/variant in the text.\n",
    "* Feature engineering:  \n",
    "  Use the wild-type/mutant residue information improve the prediction significantly. e.g., a mutation of Val-Arg may unstabilize the protein and cause loss of function.  \n",
    "  Use label_Encoder for dummy variables.\n",
    "* Modeling:  \n",
    "  Use Pipleline.FeatureUnion to combine several feature extraction into a big transformer.  \n",
    "  Using xgboost combined with model stacking improves accuracy significantly.  \n",
    "  Use TfidfVectorizer with 2 or 3-gram (4-gram is pushing my 24GB memory limit) for NLP.  \n",
    "  Use Truncated SVD for further dimensionality reduction on Tfidf matricies. \n",
    "  Due to the time constraint, I didn't get to try other NLP techniques such as lemmatization and LDA. Maybe next time.  \n",
    "\n",
    "* Acknowledgement: Thanks 1owl for his insightful kernel. My kernel is based on his framework."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/wz12/opt/anaconda3/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.\n",
      "  \"This module will be removed in 0.20.\", DeprecationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done loading!\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2\n",
       "3   3     CBL                 N454D      3\n",
       "4   4     CBL                 L399V      4"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas  as pd\n",
    "import numpy   as np\n",
    "import xgboost as xgb\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "##Load data\n",
    "train_var  = pd.read_csv('input/train_variants_stage2.csv')\n",
    "test_var   = pd.read_csv('input/stage2/stage2_test_variants.csv')\n",
    "\n",
    "trainx     = pd.read_csv('input/train_text_stage2.csv', header=None, names=[\"Text\"])\n",
    "testx      = pd.read_csv('input/stage2/stage2_test_text.csv',     sep=\"\\|\\|\", \n",
    "                         engine='python', header=None, skiprows=1, names=[\"ID\",\"Text\"])\n",
    "print(\"Done loading!\")\n",
    "train_var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3689, 4) (3689, 1) (986, 3) (986, 2)\n",
      "(3689, 5) (986, 4)\n"
     ]
    }
   ],
   "source": [
    "print(train_var.shape, trainx.shape, test_var.shape, testx.shape)\n",
    "trainx['ID'] = trainx.index\n",
    "trainx.head()\n",
    "\n",
    "#Put data and text together\n",
    "train = pd.merge(train_var, trainx, how='left', on='ID').fillna('')\n",
    "test  = pd.merge(test_var,  testx,  how='left', on='ID').fillna('')\n",
    "print(train.shape, test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3689, 4) (986, 4)\n",
      "[0 1 2 3 4 5 6 7 8]\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "#seperate training data from values\n",
    "y     = train['Class'].values - 1\n",
    "train = train.drop(['Class'], axis=1)\n",
    "pid   = test['ID'].values\n",
    "print(train.shape, test.shape)\n",
    "print(np.unique(y))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4675, 4)\n",
      "sum: 3832\n",
      "sum: 3068\n",
      "CPU times: user 1min 17s, sys: 346 ms, total: 1min 18s\n",
      "Wall time: 1min 18s\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD8CAYAAACRkhiPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFvdJREFUeJzt3X+M3HWdx/Hny22v1mJPfshk3fau/WO9XOlFOCa9XvDM\nKHrUH7E1JGRJlEYJa0KPYDRR6j9qTBP+UPHIHU1WQcqJNo1I6BHwUioTz+RKbbFS2lLZSEm7lq5i\nTBnAavF9f8ynOu5u2dmZ3fnOzuf1SCbznfd8P9/5vDvia7/f+c58FRGYmVme3lD0BMzMrDgOATOz\njDkEzMwy5hAwM8uYQ8DMLGMOATOzjDkEzMwy1nQISOqT9FNJD6fHF0naJenZdH9hw7qbJY1KOirp\nmob6lZIOpufulKTZbcfMzGZiJnsCtwJHGh7fBuyOiEFgd3qMpFXAEHAZsA64S1JfGrMVuAkYTLd1\nbc3ezMzasqCZlSQtAz4IbAE+ncrrgUpa3gZUgc+l+vaIOAM8J2kUWCPpGLA0Ivakbd4HbAAefb3X\nvuSSS2LFihVNN9To5ZdfZsmSJS2N7Ta90kuv9AHupVv1Si/t9rF///5fR8Rbp1uvqRAAvg58Fnhz\nQ60UESfT8gtAKS0PAHsa1juRan9IyxPrr2vFihXs27evyWn+pWq1SqVSaWlst+mVXnqlD3Av3apX\nemm3D0nPN7PetCEg6UPAeETslzTljCIiJM3ajxBJGgaGAUqlEtVqtaXt1Gq1lsd2m17ppVf6APfS\nrXqll0710cyewFXAhyV9AHgjsFTSt4FTkvoj4qSkfmA8rT8GLG8YvyzVxtLyxPokETECjACUy+Vo\nNQ175S8C6J1eeqUPcC/dqld66VQf034wHBGbI2JZRKyg/oHvDyPio8BOYGNabSPwUFreCQxJWiRp\nJfUPgPemQ0enJa1NZwXd0DDGzMwK0OxnAlO5Hdgh6UbgeeA6gIg4JGkHcBg4C2yKiNfSmJuBe4HF\n1D8Qft0Phc3MbG7NKAQiokr9LCAi4kXg6vOst4X6mUQT6/uA1TOdpJmZzQ1/Y9jMLGMOATOzjDkE\nzMwy5hAwM8tYO2cH2Ry5Y9fPJ9WuWFjARMys53lPwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4B\nM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsY9OGgKQ3Stor6WeSDkn6Uqp/\nUdKYpAPp9oGGMZsljUo6KumahvqVkg6m5+5MF5w3M7OCNPNT0meA90RETdJC4MeSzl0g/o6I+Erj\nypJWAUPAZcDbgMckvT1dbH4rcBPwBPAIsA5fbN7MrDDT7glEXS09XJhu8TpD1gPbI+JMRDwHjAJr\nJPUDSyNiT0QEcB+wob3pm5lZO5r6TEBSn6QDwDiwKyKeSE/dIukpSfdIujDVBoDjDcNPpNpAWp5Y\nNzOzgqj+R3mTK0tvAR4EbgF+Bfya+l7Bl4H+iPiEpP8A9kTEt9OYu6kf8jkG3B4R7031fwE+FxEf\nmuJ1hoFhgFKpdOX27dtbaq5Wq3HBBRe0NLZI4y+dmVR7k/4wL3uZaL6+J1NxL92pV3ppt493v/vd\n+yOiPN16M7q8ZET8VtLjwLrGzwIkfQN4OD0cA5Y3DFuWamNpeWJ9qtcZAUYAyuVyVCqVmUzzT6rV\nKq2OLdLUl5f85bzsZaL5+p5Mxb10p17ppVN9NHN20FvTHgCSFgPvA55Jx/jP+QjwdFreCQxJWiRp\nJTAI7I2Ik8BpSWvTWUE3AA/NYi9mZjZDzewJ9APbJPVRD40dEfGwpP+SdDn1w0HHgE8CRMQhSTuA\nw8BZYFM6MwjgZuBeYDH1Q0Q+M8jMrEDThkBEPAVcMUX9Y68zZguwZYr6PmD1DOdoZmZzxN8YNjPL\nmEPAzCxjDgEzs4w5BMzMMuYQMDPLmEPAzCxjDgEzs4w5BMzMMuYQMDPLmEPAzCxjDgEzs4w5BMzM\nMuYQMDPLmEPAzCxjDgEzs4w5BMzMMuYQMDPLmEPAzCxjzVxo/o2S9kr6maRDkr6U6hdJ2iXp2XR/\nYcOYzZJGJR2VdE1D/UpJB9Nzd6YLzpuZWUGa2RM4A7wnIt4BXA6sk7QWuA3YHRGDwO70GEmrgCHg\nMmAdcFe6SD3AVuAmYDDd1s1iL2ZmNkPThkDU1dLDhekWwHpgW6pvAzak5fXA9og4ExHPAaPAGkn9\nwNKI2BMRAdzXMMbMzArQ1GcCkvokHQDGgV0R8QRQioiTaZUXgFJaHgCONww/kWoDaXli3czMCrKg\nmZUi4jXgcklvAR6UtHrC8yEpZmtSkoaBYYBSqUS1Wm1pO7VareWxRRr43ZlJtdqZP8zLXiaar+/J\nVNxLd+qVXjrVR1MhcE5E/FbS49SP5Z+S1B8RJ9OhnvG02hiwvGHYslQbS8sT61O9zggwAlAul6NS\nqcxkmn9SrVZpdWyR7tj180m1Kxb+cl72MtF8fU+m4l66U6/00qk+mjk76K1pDwBJi4H3Ac8AO4GN\nabWNwENpeScwJGmRpJXUPwDemw4dnZa0Np0VdEPDGDMzK0AzewL9wLZ0hs8bgB0R8bCk/wN2SLoR\neB64DiAiDknaARwGzgKb0uEkgJuBe4HFwKPpZmZmBZk2BCLiKeCKKeovAlefZ8wWYMsU9X3A6skj\nzMysCP7GsJlZxhwCZmYZcwiYmWVsRqeIWpP2faut4f/wwqnJxYsHm9tu+eNtvbaZ5cV7AmZmGXMI\nmJllzCFgZpYxh4CZWcYcAmZmGXMImJllzCFgZpYxh4CZWcYcAmZmGXMImJllzCFgZpYxh4CZWcYc\nAmZmGXMImJllrJkLzS+X9Likw5IOSbo11b8oaUzSgXT7QMOYzZJGJR2VdE1D/UpJB9Nzd6YLzpuZ\nWUGauZ7AWeAzEfGkpDcD+yXtSs/dERFfaVxZ0ipgCLgMeBvwmKS3p4vNbwVuAp4AHgHW4YvNm5kV\nZto9gYg4GRFPpuWXgCPAwOsMWQ9sj4gzEfEcMAqskdQPLI2IPRERwH3AhrY7MDOzls3oMwFJK4Ar\nqP8lD3CLpKck3SPpwlQbAI43DDuRagNpeWLdzMwK0vTlJSVdADwAfCoiTkvaCnwZiHT/VeATszEp\nScPAMECpVKJarba0nVqt1vLYtryytK3hry5ZPKmms31Ux5vYbhH9zkBh78kccC/dqVd66VQfTYWA\npIXUA+D+iPg+QEScanj+G8DD6eEYsLxh+LJUG0vLE+uTRMQIMAJQLpejUqk0M81JqtUqrY5tS5vX\nGH7syORrDC+4eJDKpaenH1y+tq3XnmuFvSdzwL10p17ppVN9NHN2kIC7gSMR8bWGen/Dah8Bnk7L\nO4EhSYskrQQGgb0RcRI4LWlt2uYNwEOz1IeZmbWgmT2Bq4CPAQclHUi1zwPXS7qc+uGgY8AnASLi\nkKQdwGHqZxZtSmcGAdwM3Asspn5WkM8MMjMr0LQhEBE/BqY6n/+R1xmzBdgyRX0fsHomEzQzs7nj\nbwybmWXMIWBmljGHgJlZxhwCZmYZcwiYmWXMIWBmljGHgJlZxhwCZmYZcwiYmWXMIWBmljGHgJlZ\nxhwCZmYZcwiYmWXMIWBmljGHgJlZxhwCZmYZcwiYmWXMIWBmlrFmLjS/XNLjkg5LOiTp1lS/SNIu\nSc+m+wsbxmyWNCrpqKRrGupXSjqYnrszXXDezMwK0syewFngMxGxClgLbJK0CrgN2B0Rg8Du9Jj0\n3BBwGbAOuEtSX9rWVuAmYDDd1s1iL2ZmNkPThkBEnIyIJ9PyS8ARYABYD2xLq20DNqTl9cD2iDgT\nEc8Bo8AaSf3A0ojYExEB3NcwxszMCjCjzwQkrQCuAJ4AShFxMj31AlBKywPA8YZhJ1JtIC1PrJuZ\nWUEWNLuipAuAB4BPRcTpxsP5ERGSYrYmJWkYGAYolUpUq9WWtlOr1Voe25ZXlrY1/NUliyfVdLaP\n6ngT2y2i3xko7D2ZA+6lO/VKL53qo6kQkLSQegDcHxHfT+VTkvoj4mQ61DOe6mPA8obhy1JtLC1P\nrE8SESPACEC5XI5KpdJcNxNUq1VaHduWfd9qa/hjR05Nqi24eJDKpaenH1y+tq3XnmuFvSdzwL10\np17ppVN9NHN2kIC7gSMR8bWGp3YCG9PyRuChhvqQpEWSVlL/AHhvOnR0WtLatM0bGsaYmVkBmtkT\nuAr4GHBQ0oFU+zxwO7BD0o3A88B1ABFxSNIO4DD1M4s2RcRradzNwL3AYuDRdDMzs4JMGwIR8WPg\nfOfzX32eMVuALVPU9wGrZzJBMzObO/7GsJlZxhwCZmYZcwiYmWXMIWBmljGHgJlZxhwCZmYZcwiY\nmWXMIWBmljGHgJlZxpr+FdF56ZUX2/4xNzOzXuY9ATOzjDkEzMwy5hAwM8uYQ8DMLGMOATOzjDkE\nzMwy5hAwM8uYQ8DMLGPNXGj+Hknjkp5uqH1R0pikA+n2gYbnNksalXRU0jUN9SslHUzP3ZkuNm9m\nZgVqZk/gXmDdFPU7IuLydHsEQNIqYAi4LI25S1JfWn8rcBMwmG5TbdPMzDpo2hCIiB8Bv2lye+uB\n7RFxJiKeA0aBNZL6gaURsSciArgP2NDqpM3MbHa085nALZKeSoeLLky1AeB4wzonUm0gLU+sm5lZ\ngVr9AbmtwJeBSPdfBT4xW5OSNAwMA5RKJarVakvbqZ3tozq+dLam1TGvLlk8qaZme2nx36pTarVa\ny+9nt3Ev3alXeulUHy2FQEScOrcs6RvAw+nhGLC8YdVlqTaWlifWz7f9EWAEoFwuR6VSaWWaVB95\ngMqlp1saW6THjpyaVFtw8WBzvZSvnYMZzZ5qtUqr72e3cS/dqVd66VQfLR0OSsf4z/kIcO7MoZ3A\nkKRFklZS/wB4b0ScBE5LWpvOCroBeKiNeZuZ2SyYdk9A0neBCnCJpBPAF4CKpMupHw46BnwSICIO\nSdoBHAbOApsi4rW0qZupn2m0GHg03czMrEDThkBEXD9F+e7XWX8LsGWK+j5g9YxmZ2Zmc8rfGDYz\ny5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTM\nzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsY9OGgKR7JI1LerqhdpGkXZKeTfcXNjy3\nWdKopKOSrmmoXynpYHruznTBeTMzK1AzewL3Ausm1G4DdkfEILA7PUbSKmAIuCyNuUtSXxqzFbgJ\nGEy3ids0M7MOmzYEIuJHwG8mlNcD29LyNmBDQ317RJyJiOeAUWCNpH5gaUTsiYgA7msYY2ZmBWn1\nM4FSRJxMyy8ApbQ8ABxvWO9Eqg2k5Yl1MzMr0IJ2NxARISlmYzLnSBoGhgFKpRLVarWl7dTO9lEd\nXzqLM+uMV5csnlRTs720+G/VKbVareX3s9u4l+7UK710qo9WQ+CUpP6IOJkO9Yyn+hiwvGG9Zak2\nlpYn1qcUESPACEC5XI5KpdLSJKuPPEDl0tMtjS3SY0dOTaotuHiwuV7K187BjGZPtVql1fez27iX\n7tQrvXSqj1YPB+0ENqbljcBDDfUhSYskraT+AfDedOjotKS16aygGxrGmJlZQabdE5D0XaACXCLp\nBPAF4HZgh6QbgeeB6wAi4pCkHcBh4CywKSJeS5u6mfqZRouBR9PNzMwKNG0IRMT153nq6vOsvwXY\nMkV9H7B6RrMzM7M55W8Mm5llzCFgZpYxh4CZWcba/p5AN3vpd3+YdLrle/++dJ61zczy4z0BM7OM\nOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMws\nYw4BM7OMOQTMzDLWVghIOibpoKQDkval2kWSdkl6Nt1f2LD+Zkmjko5KuqbdyZuZWXtmY0/g3RFx\neUSU0+PbgN0RMQjsTo+RtAoYAi4D1gF3Seqbhdc3M7MWzcXhoPXAtrS8DdjQUN8eEWci4jlgFFgz\nB69vZmZNajcEAnhM0n5Jw6lWioiTafkF4NylvAaA4w1jT6SamZkVpN3LS74zIsYkXQrskvRM45MR\nEZJiphtNgTIMUCqVqFarLU3uj29YxKtLVvxFrTq+sKVtddKrSxZPqulsH9XxpdMPbvHfqlNqtVrL\n72e3cS/dqVd66VQfbYVARIyl+3FJD1I/vHNKUn9EnJTUD4yn1ceA5Q3Dl6XaVNsdAUYAyuVyVCqV\nlub339/fzuKXj/1FrfI33X+N4YnXRQZYcPEglUtPTz+4fO0czGj2VKtVWn0/u4176U690kun+mj5\ncJCkJZLefG4Z+FfgaWAnsDGtthF4KC3vBIYkLZK0EhgE9rb6+mZm1r529gRKwIOSzm3nOxHxA0k/\nAXZIuhF4HrgOICIOSdoBHAbOApsi4rW2Zm9mZm1pOQQi4hfAO6aovwhcfZ4xW4Atrb6mmZnNLn9j\n2MwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMOQTMzDLm\nEDAzy5hDwMwsYw4BM7OMOQTMzDLmEDAzy5hDwMwsYw4BM7OMtXON4ZZIWgf8O9AHfDMibu/0HGwO\n7PtWc+u9srT5dZtV/vjsbs8sIx3dE5DUB/wn8H5gFXC9pFWdnIOZmf1Zpw8HrQFGI+IXEfF7YDuw\nvsNzMDOzpNOHgwaA4w2PTwD/1OE5mM2OV16c/UNbzfIhMJslHf9MoBmShoHh9LAm6WiLm7oE+PXs\nzKpwTfbyiTmfSJvm4D0prOcC//c16z1n+N9K12u3j79tZqVOh8AYsLzh8bJU+wsRMQKMtPtikvZF\nRLnd7XSDXumlV/oA99KteqWXTvXR6c8EfgIMSlop6a+AIWBnh+dgZmZJR/cEIuKspH8D/of6KaL3\nRMShTs7BzMz+rOOfCUTEI8AjHXq5tg8pdZFe6aVX+gD30q16pZeO9KGI6MTrmJlZF/LPRpiZZawn\nQ0DSOklHJY1Kuq3o+bRD0j2SxiU9XfRc2iFpuaTHJR2WdEjSrUXPqVWS3ihpr6SfpV6+VPSc2iGp\nT9JPJT1c9FzaIemYpIOSDkjaV/R82iHpLZK+J+kZSUck/fOcvVavHQ5KP03xc+B91L+M9hPg+og4\nXOjEWiTpXUANuC8iVhc9n1ZJ6gf6I+JJSW8G9gMb5uP7IknAkoioSVoI/Bi4NSL2FDy1lkj6NFAG\nlkbEh4qeT6skHQPKETHvvyMgaRvwvxHxzXQm5Zsi4rdz8Vq9uCfQUz9NERE/An5T9DzaFREnI+LJ\ntPwScIT6N8jnnairpYcL021e/jUlaRnwQeCbRc/F6iT9NfAu4G6AiPj9XAUA9GYITPXTFPPy/2x6\nlaQVwBXAE8XOpHXpEMoBYBzYFRHztZevA58F/lj0RGZBAI9J2p9+dWC+Wgn8CvhWOkz3TUlL5urF\nejEErItJugB4APhURJwuej6tiojXIuJy6t96XyNp3h2qk/QhYDwi9hc9l1nyzvSevB/YlA6lzkcL\ngH8EtkbEFcDLwJx9ttmLIdDUT1NY56Xj5w8A90fE94uez2xIu+mPA+uKnksLrgI+nI6lbwfeI+nb\nxU6pdRExlu7HgQepHxqej04AJxr2Lr9HPRTmRC+GgH+aogulD1PvBo5ExNeKnk87JL1V0lvS8mLq\nJyE8U+ysZi4iNkfEsohYQf2/kx9GxEcLnlZLJC1JJxyQDp38KzAvz6iLiBeA45L+LpWuBubsBIqu\n/BXRdvTaT1NI+i5QAS6RdAL4QkTcXeysWnIV8DHgYDqWDvD59A3y+aYf2JbORHsDsCMi5vXplT2g\nBDxY/1uDBcB3IuIHxU6pLbcA96c/ZH8BzNlvh/fcKaJmZta8XjwcZGZmTXIImJllzCFgZpYxh4CZ\nWcYcAmZmGXMImJllzCFgZpYxh4CZWcb+H/6hz2+6J5nXAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11b14aef0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%%time\n",
    "##########################\n",
    "####Feature Engineering 1\n",
    "##########################\n",
    "%matplotlib inline\n",
    "import re\n",
    "df_all= pd.concat((train, test), axis=0, ignore_index=True)\n",
    "print(df_all.shape)\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#delimiters = [\" \", \"(\", \")\", \"/\", \".\", \",\"]\n",
    "#regex = '|'.join(map(re.escape, delimiters))\n",
    "regex = ' |\\)|\\.\\(|\\/|-|\\[|\\]||{|\\}'\n",
    "#print(\"regex:\", regex)\n",
    "df_all['Gene_Share'] = df_all.apply(\n",
    "    lambda r:sum([1 for w in r['Gene'].split(' ') \\\n",
    "                  if w.lower() in re.split(regex, r.Text.lower()) ]), axis=1)\n",
    "df_all['Variation_Share'] = df_all.apply(\n",
    "    lambda r:sum([1 for w in r['Variation'].split(' ') \\\n",
    "                  if w.lower() in re.split(regex, r.Text.lower()) ]), axis=1)\n",
    "print(\"sum:\", sum(df_all['Gene_Share']))\n",
    "print(\"sum:\", sum(df_all['Variation_Share']))\n",
    "df_all['Gene_Share'].hist(alpha=0.5) #check whether the gene name is mentioned in the text, split by ' '\n",
    "df_all['Variation_Share'].hist(alpha=0.4) #Variants may contain up to 5 words separated by space"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene found in corresponding training text,     sum:3371, ratio:91.4%\n",
      "Varaints found in corresponding training text, sum:2347, ratio:63.6%\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "      <th>Gene_Share</th>\n",
       "      <th>Variation_Share</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3646</th>\n",
       "      <td>3646</td>\n",
       "      <td>RUNX1</td>\n",
       "      <td>S291fsX300</td>\n",
       "      <td>Introduction  Myelodysplastic syndromes (MDS) ...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3674</th>\n",
       "      <td>3674</td>\n",
       "      <td>CCND3</td>\n",
       "      <td>D286_L292trunc</td>\n",
       "      <td>The activities of cyclin D-dependent kinases s...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3681</th>\n",
       "      <td>3681</td>\n",
       "      <td>NFE2L2</td>\n",
       "      <td>D29N</td>\n",
       "      <td>The Nrf2 (nuclear factor erythroid 2 [NF-E2]-r...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3683</th>\n",
       "      <td>3683</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>L115P</td>\n",
       "      <td>Genetic alterations that activate the mitogen-...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3686</th>\n",
       "      <td>3686</td>\n",
       "      <td>MAP2K1</td>\n",
       "      <td>H119P</td>\n",
       "      <td>Genetic alterations that activate the mitogen-...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID    Gene       Variation  \\\n",
       "3646  3646   RUNX1      S291fsX300   \n",
       "3674  3674   CCND3  D286_L292trunc   \n",
       "3681  3681  NFE2L2            D29N   \n",
       "3683  3683  MAP2K1           L115P   \n",
       "3686  3686  MAP2K1           H119P   \n",
       "\n",
       "                                                   Text  Gene_Share  \\\n",
       "3646  Introduction  Myelodysplastic syndromes (MDS) ...           0   \n",
       "3674  The activities of cyclin D-dependent kinases s...           0   \n",
       "3681  The Nrf2 (nuclear factor erythroid 2 [NF-E2]-r...           0   \n",
       "3683  Genetic alterations that activate the mitogen-...           0   \n",
       "3686  Genetic alterations that activate the mitogen-...           0   \n",
       "\n",
       "      Variation_Share  \n",
       "3646                1  \n",
       "3674                0  \n",
       "3681                0  \n",
       "3683                1  \n",
       "3686                0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD8CAYAAACRkhiPAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAFVZJREFUeJzt3W+MXXed3/H3BzsNXgeXQGBkHKvJA2/VJNUmzchNxe5q\nssDGS1EThISMVBIBwkhJKWhX6hKeAEKWeLALbbolkiEQp/yxrA1RIpRs5aQZUaQmxqEBxzYBiwTF\nXideoNQMsF7s/fbBnOxexpPM9b0zc2fu7/2Srubc7z2/c3/f3NifOb977nWqCklSm14x6glIkkbH\nEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWELhkCSVybZn+Q7SQ4l+URX/3iS40me7G5v\n7Rlze5KjSZ5OckNP/dokB7vH7kiSpWlLktSPLPSJ4e4v6vVVNZPkAuCbwIeAbcBMVf3ZnP2vAL4K\nbAXeADwM/HZVnU2yH/iPwOPAg8AdVfXQyz3/JZdcUpdddtkgvfGLX/yC9evXDzR2pRmXXsalD7CX\nlWpcehm2jyeeeOLHVfW6hfZbu9AONZsSM93dC7rbyyXHjcCeqjoNPJPkKLA1ybPAhqp6DCDJPcBN\nwMuGwGWXXcaBAwcWmua8pqenmZqaGmjsSjMuvYxLH2AvK9W49DJsH0l+1M9+fb0nkGRNkieBk8C+\nqnq8e+iDSb6b5AtJLu5qm4DneoYf62qbuu25dUnSiCy4HPQbOyevBu4DPgj8DfBjZs8KPglsrKr3\nJvkL4LGq+lI35i5mf9t/FvhUVb25q/8e8KdV9bZ5nmcHsANgYmLi2j179gzU3MzMDBdddNFAY1ea\ncellXPoAe1mpxqWXYfu4/vrrn6iqyYX2W3A5qFdV/SzJo8C23vcCknwO+Hp39ziwuWfYpV3teLc9\ntz7f8+wCdgFMTk7WoKdE43JaCOPTy7j0AfayUo1LL8vVRz9XB72uOwMgyTrgLcD3kmzs2e3twFPd\n9gPA9iQXJrkc2ALsr6oTwKkk13VvNt8M3L+IvUiSzlM/ZwIbgd1J1jAbGnur6utJ/nuSq5ldDnoW\n+ABAVR1Kshc4DJwBbquqs92xbgXuBtYxu0T0sm8KS5KWVj9XB30XuGae+rtfZsxOYOc89QPAVec5\nR0nSEvETw5LUMENAkhpmCEhSw87rElEtj8/s+/45tWsuGMFEJI09zwQkqWGGgCQ1zBCQpIYZApLU\nMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0z\nBCSpYYaAJDVswRBI8sok+5N8J8mhJJ/o6q9Jsi/JD7qfF/eMuT3J0SRPJ7mhp35tkoPdY3ckydK0\nJUnqRz9nAqeBP6iq3wGuBrYluQ74CPBIVW0BHunuk+QKYDtwJbAN+GySNd2x7gTeD2zpbtsWsRdJ\n0nlaMARq1kx394LuVsCNwO6uvhu4qdu+EdhTVaer6hngKLA1yUZgQ1U9VlUF3NMzRpI0An29J5Bk\nTZIngZPAvqp6HJioqhPdLs8DE932JuC5nuHHutqmbntuXZI0Imv72amqzgJXJ3k1cF+Sq+Y8Xklq\nsSaVZAewA2BiYoLp6emBjjMzMzPw2FHa9Lenz6nNnP71quxlrtX6mszHXlamcellufroKwReVFU/\nS/Ios2v5LyTZWFUnuqWek91ux4HNPcMu7WrHu+259fmeZxewC2BycrKmpqbOZ5r/YHp6mkHHjtJn\n9n3/nNo1F/z1quxlrtX6mszHXlamcellufro5+qg13VnACRZB7wF+B7wAHBLt9stwP3d9gPA9iQX\nJrmc2TeA93dLR6eSXNddFXRzzxhJ0gj0cyawEdjdXeHzCmBvVX09yf8G9iZ5H/Aj4J0AVXUoyV7g\nMHAGuK1bTgK4FbgbWAc81N0kSSOyYAhU1XeBa+ap/wR400uM2QnsnKd+ALjq3BGSpFHwE8OS1DBD\nQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQk\nqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhC4ZAks1JHk1yOMmhJB/q6h9P\ncjzJk93trT1jbk9yNMnTSW7oqV+b5GD32B1JsjRtSZL6sbaPfc4Af1JV307yKuCJJPu6xz5TVX/W\nu3OSK4DtwJXAG4CHk/x2VZ0F7gTeDzwOPAhsAx5anFYkSedrwTOBqjpRVd/utn8OHAE2vcyQG4E9\nVXW6qp4BjgJbk2wENlTVY1VVwD3ATUN3IEkaWGb/Pu5z5+Qy4BvAVcAfA+8B/h9wgNmzhf+b5C+A\nx6rqS92Yu5j9bf9Z4FNV9eau/nvAn1bV2+Z5nh3ADoCJiYlr9+zZM1BzMzMzXHTRRQONHaWTPz99\nTu238utV2ctcq/U1mY+9rEzj0suwfVx//fVPVNXkQvv1sxwEQJKLgHuBD1fVqSR3Ap8Eqvv558B7\nB5zvb6iqXcAugMnJyZqamhroONPT0ww6dpQ+s+/759SuueCvV2Uvc63W12Q+9rIyjUsvy9VHX1cH\nJbmA2QD4clV9DaCqXqiqs1X198DngK3d7seBzT3DL+1qx7vtuXVJ0oj0c3VQgLuAI1X16Z76xp7d\n3g481W0/AGxPcmGSy4EtwP6qOgGcSnJdd8ybgfsXqQ9J0gD6WQ56I/Bu4GCSJ7vaR4F3Jbma2eWg\nZ4EPAFTVoSR7gcPMXll0W3dlEMCtwN3AOmbfJ/DKIEkaoQVDoKq+Ccx3Pf+DLzNmJ7BznvoBZt9U\nliStAH5iWJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapgh\nIEkNMwQkqWGGgCQ1zBCQpIYZApLUsL7/oXmdhwNfHGr4v3z+hXOLr93S33En3zPUc0tqi2cCktQw\nQ0CSGmYISFLDDAFJatiCIZBkc5JHkxxOcijJh7r6a5LsS/KD7ufFPWNuT3I0ydNJbuipX5vkYPfY\nHUmyNG1JkvrRz5nAGeBPquoK4DrgtiRXAB8BHqmqLcAj3X26x7YDVwLbgM8mWdMd607g/cCW7rZt\nEXuRJJ2nBUOgqk5U1be77Z8DR4BNwI3A7m633cBN3faNwJ6qOl1VzwBHga1JNgIbquqxqirgnp4x\nkqQROK/3BJJcBlwDPA5MVNWJ7qHngYluexPwXM+wY11tU7c9ty5JGpG+PyyW5CLgXuDDVXWqdzm/\nqipJLdakkuwAdgBMTEwwPT090HFmZmYGHjuUX24Yaviv1q87p5Yza5g+2cdxR9HveRjZa7IE7GVl\nGpdelquPvkIgyQXMBsCXq+prXfmFJBur6kS31HOyqx8HNvcMv7SrHe+259bPUVW7gF0Ak5OTNTU1\n1V83c0xPTzPo2KEM+Ynhh4+c+4nhta/dwtTrTy08ePIdQz33UhvZa7IE7GVlGpdelquPfq4OCnAX\ncKSqPt3z0APALd32LcD9PfXtSS5McjmzbwDv75aOTiW5rjvmzT1jJEkj0M+ZwBuBdwMHkzzZ1T4K\nfArYm+R9wI+AdwJU1aEke4HDzF5ZdFtVne3G3QrcDawDHupukqQRWTAEquqbwEtdz/+mlxizE9g5\nT/0AcNX5TFCStHT8xLAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCk\nhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWrY\ngiGQ5AtJTiZ5qqf28STHkzzZ3d7a89jtSY4meTrJDT31a5Mc7B67I0kWvx1J0vno50zgbmDbPPXP\nVNXV3e1BgCRXANuBK7sxn02yptv/TuD9wJbuNt8xJUnLaMEQqKpvAD/t83g3Anuq6nRVPQMcBbYm\n2QhsqKrHqqqAe4CbBp20JGlxDPOewAeTfLdbLrq4q20CnuvZ51hX29Rtz61LkkZo7YDj7gQ+CVT3\n88+B9y7WpJLsAHYATExMMD09PdBxZmZmBh47lF9uGGr4r9avO6eWM2uYPtnHcUfR73kY2WuyBOxl\nZRqXXparj4FCoKpeeHE7yeeAr3d3jwObe3a9tKsd77bn1l/q+LuAXQCTk5M1NTU1yDSZnp5m0LFD\nOfDFoYY/fOSFc2prX7uFqdefWnjw5DuGeu6lNrLXZAnYy8o0Lr0sVx8DhUCSjVV1orv7duDFK4ce\nAL6S5NPAG5h9A3h/VZ1NcirJdcDjwM3Afx1u6n345U+G/gtZksbZgiGQ5KvAFHBJkmPAx4CpJFcz\nuxz0LPABgKo6lGQvcBg4A9xWVWe7Q93K7JVG64CHupskaYQWDIGqetc85bteZv+dwM556geAq85r\ndpKkJeUnhiWpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZ\nApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMWDIEkX0hy\nMslTPbXXJNmX5Afdz4t7Hrs9ydEkTye5oad+bZKD3WN3JMnityNJOh/9nAncDWybU/sI8EhVbQEe\n6e6T5ApgO3BlN+azSdZ0Y+4E3g9s6W5zjylJWmYLhkBVfQP46ZzyjcDubns3cFNPfU9Vna6qZ4Cj\nwNYkG4ENVfVYVRVwT88YSdKIrB1w3ERVnei2nwcmuu1NwGM9+x3rar/utufW55VkB7ADYGJigunp\n6YEmOXNmDdMnNww0dpR+tX7dObX028uA/62Wy8zMzMCv50pjLyvTuPSyXH0MGgL/oKoqSS3GZHqO\nuQvYBTA5OVlTU1MDHWf6wXuZev2pRZzZ8nj4yAvn1Na+dkt/vUy+YwlmtHimp6cZ9PVcaexlZRqX\nXparj0GvDnqhW+Kh+3myqx8HNvfsd2lXO95tz61LkkZo0BB4ALil274FuL+nvj3JhUkuZ/YN4P3d\n0tGpJNd1VwXd3DNGkjQiCy4HJfkqMAVckuQY8DHgU8DeJO8DfgS8E6CqDiXZCxwGzgC3VdXZ7lC3\nMnul0Trgoe4mSRqhBUOgqt71Eg+96SX23wnsnKd+ALjqvGYnSVpSfmJYkhpmCEhSwwwBSWqYISBJ\nDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQw\nQ0CSGmYISFLDDAFJapghIEkNMwQkqWFDhUCSZ5McTPJkkgNd7TVJ9iX5Qffz4p79b09yNMnTSW4Y\ndvKSpOEsxpnA9VV1dVVNdvc/AjxSVVuAR7r7JLkC2A5cCWwDPptkzSI8vyRpQEuxHHQjsLvb3g3c\n1FPfU1Wnq+oZ4CiwdQmeX5LUp2FDoICHkzyRZEdXm6iqE93288BEt70JeK5n7LGuJkkakbVDjv/d\nqjqe5PXAviTf632wqipJne9Bu0DZATAxMcH09PRAk5s5s4bpkxsGGjtKv1q/7pxa+u1lwP9Wy2Vm\nZmbg13OlsZeVaVx6Wa4+hgqBqjre/TyZ5D5ml3deSLKxqk4k2Qic7HY/DmzuGX5pV5vvuLuAXQCT\nk5M1NTU10PymH7yXqdefGmjsKD185IVzamtfu6W/XibfsQQzWjzT09MM+nquNPayMo1LL8vVx8DL\nQUnWJ3nVi9vAHwJPAQ8At3S73QLc320/AGxPcmGSy4EtwP5Bn1+SNLxhzgQmgPuSvHicr1TVXyX5\nFrA3yfuAHwHvBKiqQ0n2AoeBM8BtVXV2qNlLkoYycAhU1Q+B35mn/hPgTS8xZiewc9DnlCQtLj8x\nLEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkho27NdGrGg//9tfn/Pp2zf/i4mX2FuS2uOZgCQ1zBCQ\npIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlq\nmCEgSQ1b9n9PIMk24L8Aa4DPV9WnlnsOWgIHvtjffr/c0P++/Zp8z+IeT2rIsp4JJFkD/Dfgj4Ar\ngHcluWI55yBJ+kfLvRy0FThaVT+sqr8D9gA3LvMcJEmd5V4O2gQ813P/GPCvl3kO0uL45U8Wf2mr\nXy6BaZGsyH9jOMkOYEd3dybJ0wMe6hLgx4szq5Hrs5f3LvlEhrQEr8nIeh7h/1+L3nODf1ZWvGH7\n+Gf97LTcIXAc2Nxz/9Ku9huqahewa9gnS3KgqiaHPc5KMC69jEsfYC8r1bj0slx9LPd7At8CtiS5\nPMk/AbYDDyzzHCRJnWU9E6iqM0n+A/A/mL1E9AtVdWg55yBJ+kfL/p5AVT0IPLhMTzf0ktIKMi69\njEsfYC8r1bj0six9pKqW43kkSSuQXxshSQ0byxBIsi3J00mOJvnIqOczjCRfSHIyyVOjnsswkmxO\n8miSw0kOJfnQqOc0qCSvTLI/yXe6Xj4x6jkNI8maJP8nyddHPZdhJHk2ycEkTyY5MOr5DCPJq5P8\nZZLvJTmS5N8s2XON23JQ99UU3wfewuyH0b4FvKuqDo90YgNK8vvADHBPVV016vkMKslGYGNVfTvJ\nq4AngJtW4+uSJMD6qppJcgHwTeBDVfXYiKc2kCR/DEwCG6rqbaOez6CSPAtMVtWq/4xAkt3A/6qq\nz3dXUv5WVf1sKZ5rHM8ExuqrKarqG8BPRz2PYVXViar6drf9c+AIs58gX3Vq1kx394Lutip/m0py\nKfBvgc+Pei6aleSfAr8P3AVQVX+3VAEA4xkC8301xar8y2ZcJbkMuAZ4fLQzGVy3hPIkcBLYV1Wr\ntZf/DPwn4O9HPZFFUMDDSZ7ovnVgtboc+Bvgi90y3eeTrF+qJxvHENAKluQi4F7gw1V1atTzGVRV\nna2qq5n91PvWJKtuqS7J24CTVfXEqOeySH63e03+CLitW0pdjdYC/wq4s6quAX4BLNl7m+MYAn19\nNYWWX7d+fi/w5ar62qjnsxi60/RHgW2jnssA3gj8u24tfQ/wB0m+NNopDa6qjnc/TwL3Mbs0vBod\nA471nF3+JbOhsCTGMQT8aooVqHsz9S7gSFV9etTzGUaS1yV5dbe9jtmLEL432lmdv6q6vaourarL\nmP1z8j+r6t+PeFoDSbK+u+CAbunkD4FVeUVdVT0PPJfkn3elNwFLdgHFivwW0WGM21dTJPkqMAVc\nkuQY8LGqumu0sxrIG4F3Awe7tXSAj3afIF9tNgK7uyvRXgHsrapVfXnlGJgA7pv9XYO1wFeq6q9G\nO6WhfBD4cveL7A+BJfvu8LG7RFSS1L9xXA6SJPXJEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkN\nMwQkqWH/H5pLHUzi7DLjAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11b2c8ef0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#investigate why some genes/variants are not found in texts\n",
    "tmp = df_all.loc[range(len(train))]\n",
    "print(\"Gene found in corresponding training text,     sum:{}, ratio:{}%\".format(sum(tmp.Gene_Share), \n",
    "                            round(100*sum(tmp.Gene_Share)/len(tmp),1)))\n",
    "print(\"Varaints found in corresponding training text, sum:{}, ratio:{}%\".format(sum(tmp.Variation_Share), \n",
    "                            round(100*sum(tmp.Variation_Share)/len(tmp),1)))\n",
    "\n",
    "tmp.Gene_Share.hist(alpha=0.5) #check whether the gene name is mentioned in the text, split by ' '\n",
    "tmp.Variation_Share.hist(alpha=0.4) #Variants may contain up to 5 words separated by space\n",
    "gene_not_share = tmp.loc[tmp.Gene_Share == 0]\n",
    "vari_not_share = tmp.loc[tmp.Variation_Share == 0]\n",
    "gene_not_share.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "318"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(gene_not_share)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "188                                      Exon 19 deletion\n",
       "197                                     Exon 20 insertion\n",
       "198                                     Exon 19 insertion\n",
       "412                          DNA binding domain deletions\n",
       "428                         DNA binding domain insertions\n",
       "480                 DNA binding domain missense mutations\n",
       "631                                      Copy Number Loss\n",
       "743                                    Exon 20 insertions\n",
       "1410                              FGFR3 - BAIAP2L1 Fusion\n",
       "1640                    FLT3 internal tandem duplications\n",
       "3008                                    Exon 11 mutations\n",
       "3025                                     Exon 9 mutations\n",
       "3060                                     Exon 1 mutations\n",
       "3062                                     Exon 2 mutations\n",
       "3073              Truncating Mutations in the PEST Domain\n",
       "3078    Truncating Mutations Upstream of Transactivati...\n",
       "3221                               TRKAIII Splice Variant\n",
       "3389                       Single Nucleotide Polymorphism\n",
       "3616                                    Exon 11 deletions\n",
       "3947                       Single Nucleotide Polymorphism\n",
       "4580                                    Exon 11 deletions\n",
       "Name: Variation, dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.loc[df_all['Variation_Share']>=3, 'Variation']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9 55\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAADHdJREFUeJzt3XGsnfVdx/H3Z3S4ge7OABpswYuBkDX+IXrDVJQY2Uxr\nV1gWozTqH4asYsbC9A/TGf/xP5YYY5YQlwaQJW4QhM1QWwcaN9EEkZZhBuswFbtRnLZzeufUhDG/\n/nGP8aZyT8/lntPnnO/er6Rpz3POPc8nN7mf+5zv83uepqqQJPX1hqEDSJJmy6KXpOYseklqzqKX\npOYseklqzqKXpOYseklqzqKXpOYseklqbtvQAQAuvfTSWl5eHjqGJC2UY8eOfaWqLjvX6+ai6JeX\nlzl69OjQMSRpoST54iSvc3QjSc1Z9JLUnEUvSc1Z9JLU3KBFn2RvkoOrq6tDxpCk1gYt+qo6VFX7\nl5aWhowhSa05upGk5ix6SWpuLi6Y0vxYPnB4w+dO3rXnPCaRNC0e0UtScxa9JDVn0UtScxa9JDVn\n0UtSc666UQuuFpI25hG9JDVn0UtScxa9JDVn0UtScxa9JDVn0UtScxa9JDVn0UtScxa9JDU39aJP\n8rYkH0nycJJfmfb7S5I2Z6KiT3JfktNJnjtr+64kLyQ5keQAQFUdr6rbgZ8Fbph+ZEnSZkx6RH8/\nsGv9hiQXAHcDu4GdwL4kO0fP3QwcBo5MLakk6XWZqOir6gngq2dtvh44UVUvVtUrwIPALaPXP1pV\nu4Gfn2ZYSdLmbeXulduBl9Y9PgW8PclPAO8Bvo0xR/RJ9gP7Aa688sotxJAkjTP12xRX1WeAz0zw\nuoPAQYCVlZWadg5J0pqtrLp5Gbhi3eMdo22SpDmylaJ/GrgmyVVJLgRuBR7dzBsk2Zvk4Orq6hZi\nSJLGmXR55QPAk8C1SU4lua2qXgXuAB4DjgMPVdXzm9l5VR2qqv1LS0ubzS1JmtBEM/qq2rfB9iO4\nhFKS5pq3QJCk5gYtemf0kjR7gxa9M3pJmj1HN5LUnEUvSc1N/cpYSf9n+cDhDZ87edee85hE38o8\nGStJzXkyVpKac0YvSc1Z9JLUnEUvSc15MlaSmvNkrCQ15+hGkpqz6CWpOa+MnZFxV0SCV0VKOn88\nopek5lx1I0nNuepGkppzdCNJzVn0ktScRS9JzVn0ktScRS9Jzbm8UpKac3mlJDXn6EaSmrPoJak5\ni16SmrPoJak5i16SmrPoJak5i16SmrPoJak5r4yVpOa8MlaSmnN0I0nNWfSS1JxFL0nNWfSS1JxF\nL0nNWfSS1JxFL0nNWfSS1JxFL0nNWfSS1JxFL0nNWfSS1Jx3r5Sk5rx7pSQ15+hGkpqz6CWpOYte\nkpqz6CWpOYtekpqz6CWpOYtekpqz6CWpOYtekpqz6CWpOYtekpqz6CWpOYtekpqz6CWpOYtekpqz\n6CWpOYtekprbNu03TPJuYA/wFuDeqnp82vuQJE1uoiP6JPclOZ3kubO270ryQpITSQ4AVNUfVdV7\ngduBn5t+ZEnSZkw6urkf2LV+Q5ILgLuB3cBOYF+Snete8puj5yVJA5qo6KvqCeCrZ22+HjhRVS9W\n1SvAg8AtWfMh4E+q6pnpxpUkbdZWTsZuB15a9/jUaNv7gXcAP5Pk9o2+OMn+JEeTHD1z5swWYkiS\nxpn6ydiq+jDw4QledxA4CLCyslLTziFJWrOVI/qXgSvWPd4x2iZJmiNbKfqngWuSXJXkQuBW4NHp\nxJIkTcukyysfAJ4Erk1yKsltVfUqcAfwGHAceKiqnt/MzpPsTXJwdXV1s7klSROaaEZfVfs22H4E\nOPJ6d15Vh4BDKysr73297yFJGs9bIEhScxa9JDU3aNE7o5ek2Ru06KvqUFXtX1paGjKGJLU29Qum\nJC225QOHxz5/8q495ymJpsUZvSQ154xekppzRi9JzTm6kaTmLHpJas6il6TmPBkrSc15MlaSmnN0\nI0nNWfSS1JxFL0nNWfSS1JyrbiSpOVfdSFJzjm4kqTmLXpKas+glqTmLXpKas+glqTmLXpKacx29\nJDXnOnpJas7RjSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnNeGStJzXllrCQ15+hG\nkpqz6CWpOYtekpqz6CWpOYtekprbNnSArVo+cHjs8yfv2nOekkjSfPKIXpKas+glqTmLXpKas+gl\nqbmFPxkrSTB+Yca3+qIMb2omSc15UzNJas4ZvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1\nZ9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnMWvSQ1Z9FLUnNTL/ok35fk\n3iQPT/u9JUmbN1HRJ7kvyekkz521fVeSF5KcSHIAoKperKrbZhFWkrR5kx7R3w/sWr8hyQXA3cBu\nYCewL8nOqaaTJG3ZREVfVU8AXz1r8/XAidER/CvAg8AtU84nSdqirczotwMvrXt8Ctie5JIkHwGu\nS/LBjb44yf4kR5McPXPmzBZiSJLG2TbtN6yqfwFun+B1B4GDACsrKzXtHJKkNVs5on8ZuGLd4x2j\nbZKkObKVon8auCbJVUkuBG4FHt3MGyTZm+Tg6urqFmJIksaZdHnlA8CTwLVJTiW5rapeBe4AHgOO\nAw9V1fOb2XlVHaqq/UtLS5vNLUma0EQz+qrat8H2I8CRqSaSJE2Vt0CQpOYGLXpn9JI0e4MWvTN6\nSZo9RzeS1JxFL0nNWfSS1JwnYyWpOU/GSlJzjm4kqTmLXpKas+glqTlPxkpSc56MlaTmHN1IUnMW\nvSQ1Z9FLUnMWvSQ1N9H/MDUrSfYCe6+++uohY0jSzCwfODz2+ZN37Zl5BlfdSFJzjm4kqTmLXpKa\ns+glqTmLXpKas+glqTlvaiZJzbm8UpKac3QjSc2lqobOQJIzwBdf55dfCnxlinFmbZHyLlJWWKy8\ni5QVFivvImWFreX93qq67Fwvmoui34okR6tqZegck1qkvIuUFRYr7yJlhcXKu0hZ4fzkdXQjSc1Z\n9JLUXIeiPzh0gE1apLyLlBUWK+8iZYXFyrtIWeE85F34Gb0kabwOR/SSpDEWtuiTXJHk00k+n+T5\nJHcOnWkjSd6U5G+S/O0o628NnelcklyQ5LNJ/njoLOeS5GSSzyV5NsnRofOcS5K3Jnk4yReSHE/y\nI0Nnei1Jrh19T//3z9eSfGDoXBtJ8qujn6/nkjyQ5E1DZxonyZ2jrM/P+vu6sKObJJcDl1fVM0m+\nAzgGvLuqPj9wtP8nSYCLq+rrSd4I/BVwZ1X99cDRNpTk14AV4C1V9a6h84yT5CSwUlULsXY6yUeB\nv6yqe5JcCFxUVf82dK5xklwAvAy8vape7zUvM5NkO2s/Vzur6r+SPAQcqar7h0322pJ8P/AgcD3w\nCvAp4PaqOjGL/S3sEX1Vfbmqnhn9+9+B48D2YVO9tlrz9dHDN47+zO1v2CQ7gD3APUNn6SbJEnAj\ncC9AVb0y7yU/chPw9/NY8utsA96cZBtwEfCPA+cZ523AU1X1n1X1KvAXwHtmtbOFLfr1kiwD1wFP\nDZtkY6NRyLPAaeBPq2puswK/C/w68N9DB5lQAX+W5FiS/UOHOYergDPA749GY/ckuXjoUBO4FXhg\n6BAbqaqXgd8GvgR8GVitqseHTTXWc8CPJ7kkyUXATwNXzGpnC1/0Sb4deAT4QFV9beg8G6mqb1bV\nDwA7gOtHH93mTpJ3Aaer6tjQWTbhx0bf293A+5LcOHSgMbYBPwj8XlVdB/wHcGDYSOONxks3A384\ndJaNJPlO4BbWfpF+D3Bxkl8YNtXGquo48CHgcdbGNs8C35zV/ha66Efz7keAj1XVJ4bOM4nRx/RP\nA7uGzrKBG4CbR3PvB4GfTPIHw0Yab3Q0R1WdBj7J2txzXp0CTq37RPcwa8U/z3YDz1TVPw8dZIx3\nAP9QVWeq6hvAJ4AfHTjTWFV1b1X9UFXdCPwr8Hez2tfCFv3oBOe9wPGq+p2h84yT5LIkbx39+83A\nO4EvDJvqtVXVB6tqR1Uts/Zx/c+ram6PjJJcPDoZz2gE8lOsfSyeS1X1T8BLSa4dbboJmLsFBGfZ\nxxyPbUa+BPxwkotG3XATa+ft5laS7xr9fSVr8/mPz2pf22b1xufBDcAvAp8bzb4BfqOqjgyYaSOX\nAx8drVx4A/BQVc39ssUF8d3AJ9d+ttkGfLyqPjVspHN6P/Cx0UjkReCXBs6zodEvz3cCvzx0lnGq\n6qkkDwPPAK8Cn2X+r5B9JMklwDeA983ypPzCLq+UJE1mYUc3kqTJWPSS1JxFL0nNWfSS1JxFL0nN\nWfSS1JxFL0nNWfSS1Nz/AFT+jE6NORwqAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11b2b5b70>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAD8CAYAAAB5Pm/hAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAADGJJREFUeJzt3X+o3fddx/Hny3RVmRp1rUPyw9uRUAyiHVyywfZHHSoJ\nWZoxZDYqDAkNFSsTFMlEEAUhf4kUK+NiSwbOhlDdTGykq3Ol/lG0yZysMSuGktGEuWQW448/LJlv\n/zin7nJpbs7JOeeee9/3+YBy7/mcc8/5vPvHKx/e5/P9flJVSJL6+o55T0CSNFsGvSQ1Z9BLUnMG\nvSQ1Z9BLUnMGvSQ1Z9BLUnMGvSQ1Z9BLUnN3zXsCAPfcc08tLCzMexqStKGcP3/+m1V17+1ety6C\nfmFhgXPnzs17GpK0oST52iivs3UjSc0Z9JLUnEEvSc3NNeiTHEyydOPGjXlOQ5Jam2vQV9WZqjq6\ndevWeU5DklqzdSNJzRn0ktScQS9Jza2LC6ZmaeHYs7d87vLxA2s4E0maD1f0ktScQS9JzRn0ktSc\nQS9JzXllrCQ155WxktScrRtJas6gl6TmDHpJas6gl6TmDHpJas6gl6TmDHpJas6gl6TmDHpJas6g\nl6TmDHpJas6gl6Tmph70SX40yaeSPJPkl6f9/pKk8YwU9EmeSnItySsrxvcleTXJpSTHAKrqYlU9\nCnwM+MD0pyxJGseoK/oTwL7lA0m2AE8A+4E9wOEke4bPPQQ8C5yd2kwlSXdkpKCvqheBN1YM7wUu\nVdVrVfUmcBI4NHz96araD/zCNCcrSRrfXRP87Tbg9WWPrwDvS/Ig8FHgO1llRZ/kKHAUYOfOnRNM\nQ5K0mkmC/m1V1QvACyO8bglYAlhcXKxpz0OSNDDJrpurwI5lj7cPxyRJ68gkQf8ysDvJfUnuBh4G\nTo/zBh4OLkmzN+r2yqeBl4D7k1xJcqSqbgKPAc8BF4FTVXVhnA/3cHBJmr2RevRVdfgW42dxC6Uk\nrWtzvQWCrRtJmr2p77oZR1WdAc4sLi4+Mo/PXzj27KrPXz5+YI1mIkmz403NJKk5WzeS1NyGb93c\nrv0iSZudrRtJas6gl6Tm7NFLUnNzDXqvjJWk2bN1I0nNGfSS1Jw9eklqzh69JDVn60aSmjPoJak5\ng16SmjPoJak5d91IUnPuupGk5mzdSFJzBr0kNWfQS1JzBr0kNWfQS1Jzbq+UpObcXilJzdm6kaTm\nDHpJas6gl6TmDHpJas6gl6TmDHpJas6gl6TmDHpJas6gl6Tm7prnhyc5CBzctWvXPKdxSwvHnl31\n+cvHD6zRTCTpznkLBElqztaNJDVn0EtScwa9JDVn0EtScwa9JDU31+2V6sXtqNL65Ipekpoz6CWp\nOYNekpqzR7/J2EeXNh9X9JLU3ExW9Ek+AhwAvg94sqo+P4vPkSTd3sgr+iRPJbmW5JUV4/uSvJrk\nUpJjAFX1uap6BHgU+LnpTlmSNI5xWjcngH3LB5JsAZ4A9gN7gMNJ9ix7yW8Pn5ckzcnIrZuqejHJ\nworhvcClqnoNIMlJ4FCSi8Bx4K+r6ktTmmsrfikqaa1M+mXsNuD1ZY+vDMd+Ffgp4GeTPPp2f5jk\naJJzSc5dv359wmlIkm5lJl/GVtXjwOO3ec0SsASwuLhYs5iHJGnyFf1VYMeyx9uHY5KkdWLSFf3L\nwO4k9zEI+IeBnx/1j9f7mbHztFoP3/69pHGMs73yaeAl4P4kV5IcqaqbwGPAc8BF4FRVXRj1PT0z\nVpJmb5xdN4dvMX4WODu1GUmSpmqut0BIcjDJ0o0bN+Y5DUlqba5Bb+tGkmbPm5pJUnO2biSpOVs3\nktScrRtJas6gl6Tm7NFLUnP26CWpOVs3ktScQS9JzRn0ktScX8ZKUnN+GStJzdm6kaTmDHpJam4m\nh4Nr41rtCEPwGENpI/LLWElqzi9jJak5e/SS1JxBL0nNGfSS1JxBL0nNub2yodttkZS0ubiil6Tm\n3EcvSc25j16SmrN1I0nNGfSS1JxBL0nNGfSS1JxBL0nNGfSS1JxXxmosXnUrbTyu6CWpOa+MlaTm\n5tq6qaozwJnFxcVH5jmPjcb2iaRx2LqRpOYMeklqzqCXpOYMeklqzqCXpOYMeklqzqCXpOYMeklq\nzqCXpOa8qdmMePWqpPXCFb0kNWfQS1JzUw/6JO9J8mSSZ6b93pKk8Y0U9EmeSnItySsrxvcleTXJ\npSTHAKrqtao6MovJSpLGN+qK/gSwb/lAki3AE8B+YA9wOMmeqc5OkjSxkYK+ql4E3lgxvBe4NFzB\nvwmcBA5NeX6SpAlN0qPfBry+7PEVYFuSdyX5FPDeJJ+81R8nOZrkXJJz169fn2AakqTVTH0ffVX9\nG/DoCK9bApYAFhcXa9rzkCQNTLKivwrsWPZ4+3BsZJ4ZK0mzN0nQvwzsTnJfkruBh4HT47xBVZ2p\nqqNbt26dYBqSpNWMur3yaeAl4P4kV5IcqaqbwGPAc8BF4FRVXZjdVCVJd2KkHn1VHb7F+Fng7J1+\neJKDwMFdu3bd6VtIkm5jrrdAsHUjSbPnvW4kqTmDXpKam2vQu71SkmbPHr0kNWfrRpKaM+glqbm5\nnhm70ffRey7seFb7/3X5+IE1nIm0udijl6TmbN1IUnMGvSQ15z56SWrOHr0kNWfrRpKaM+glqTmD\nXpKaM+glqTl33UhSc+66kaTmbN1IUnMGvSQ1Z9BLUnMGvSQ1Z9BLUnMePKJNb5IDZDwwRRuB2ysl\nqTlbN5LUnEEvSc0Z9JLUnEEvSc0Z9JLUnEEvSc0Z9JLUnEEvSc15ZazWhUmuTr2d9Xr16u1qXq/z\n1sbjlbGS1JytG0lqzqCXpOYMeklqzqCXpOYMeklqzqCXpOYMeklqzqCXpOYMeklqzqCXpOYMeklq\nzqCXpOamfvfKJO8E/hh4E3ihqj4z7c+QJI1upBV9kqeSXEvyyorxfUleTXIpybHh8EeBZ6rqEeCh\nKc9XkjSmUVs3J4B9yweSbAGeAPYDe4DDSfYA24HXhy/71nSmKUm6UyMFfVW9CLyxYngvcKmqXquq\nN4GTwCHgCoOwH/n9JUmzM0mPfhvfXrnDIODfBzwO/FGSA8CZW/1xkqPAUYCdO3dOMA1pfuZ1StQs\nT+S6HU++Gs96OEls6l/GVtV/A780wuuWgCWAxcXFmvY8JEkDk7RWrgI7lj3ePhyTJK0jkwT9y8Du\nJPcluRt4GDg9zhskOZhk6caNGxNMQ5K0mlG3Vz4NvATcn+RKkiNVdRN4DHgOuAicqqoL43y4h4NL\n0uyN1KOvqsO3GD8LnJ3qjCRJUzXX7Y+2biRp9uYa9LZuJGn2vKBJkpqzdSNJzaVq/tcqJbkOfG3e\n85ihe4BvznsSa2Sz1LpZ6oTNU+tGrPNHqure271oXQR9d0nOVdXivOexFjZLrZulTtg8tXau0x69\nJDVn0EtScwb92lia9wTW0GapdbPUCZun1rZ12qOXpOZc0UtScwb9lL3d+bpJfjDJ80n+ZfjzB+Y5\nx2lIsiPJF5P8c5ILST4xHO9Y63cl+Yck/zSs9XeH4+1qhcExoUn+MclfDR93rfNykq8k+XKSc8Ox\nlrUa9NN3ghXn6wLHgC9U1W7gC8PHG91N4Nerag/wfuBXhmcGd6z1f4APVdVPAA8A+5K8n561AnyC\nwR1p39K1ToCfrKoHlm2rbFmrQT9ltzhf9xDw6eHvnwY+sqaTmoGq+npVfWn4+38yCIZt9Ky1quq/\nhg/fMfyvaFhrku3AAeBPlg23q3MVLWs16NfGu6vq68Pf/xV49zwnM21JFoD3An9P01qH7YwvA9eA\n56uqa61/CPwm8L/LxjrWCYN/rP8myfnhGdbQtNapnxmr1VVVJWmz1SnJ9wB/DvxaVf1Hkv9/rlOt\nVfUt4IEk3w98NsmPrXh+w9ea5MPAtao6n+TBt3tNhzqX+WBVXU3yQ8DzSb66/MlOtbqiXxvfSPLD\nAMOf1+Y8n6lI8g4GIf+ZqvqL4XDLWt9SVf8OfJHB9zDdav0A8FCSy8BJ4ENJ/pR+dQJQVVeHP68B\nnwX20rRWg35tnAY+Pvz948BfznEuU5HB0v1J4GJV/cGypzrWeu9wJU+S7wZ+GvgqzWqtqk9W1faq\nWmBwBvTfVtUv0qxOgCTvTPK9b/0O/AzwCg1rBS+Ymrrh+boPMrgT3jeA3wE+B5wCdjK4S+fHqmrl\nF7YbSpIPAn8HfIVv93N/i0GfvlutP87gi7ktDBZHp6rq95K8i2a1vmXYuvmNqvpwxzqTvIfBKh4G\nLew/q6rf71grGPSS1J6tG0lqzqCXpOYMeklqzqCXpOYMeklqzqCXpOYMeklqzqCXpOb+DxwhBATM\nAeAJAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x11b419208>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_len = [len(x) for x in df_all.Gene ]\n",
    "vari_len = [len(x) for x in df_all.Variation ]\n",
    "print(max(gene_len), max(vari_len))\n",
    "plt.figure()\n",
    "_=plt.hist(gene_len, log=True, bins=40)\n",
    "plt.figure()\n",
    "_=plt.hist(vari_len, log=True, bins=40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4675, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Text</th>\n",
       "      <th>Gene_Share</th>\n",
       "      <th>Variation_Share</th>\n",
       "      <th>Variation_0</th>\n",
       "      <th>Variation_1</th>\n",
       "      <th>Variation_2</th>\n",
       "      <th>Variation_3</th>\n",
       "      <th>...</th>\n",
       "      <th>Variation_14</th>\n",
       "      <th>Variation_15</th>\n",
       "      <th>Variation_16</th>\n",
       "      <th>Variation_17</th>\n",
       "      <th>Variation_18</th>\n",
       "      <th>Variation_19</th>\n",
       "      <th>Variation_20</th>\n",
       "      <th>Variation_21</th>\n",
       "      <th>Variation_22</th>\n",
       "      <th>var_wild_plus_mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>r</td>\n",
       "      <td>u</td>\n",
       "      <td>n</td>\n",
       "      <td>...</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>i</td>\n",
       "      <td>o</td>\n",
       "      <td>n</td>\n",
       "      <td>s</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Ts</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>W</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>W*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Q</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>QE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>N</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>L</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>LV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  \\\n",
       "0   0  FAM58A  Truncating Mutations   \n",
       "1   1     CBL                 W802*   \n",
       "2   2     CBL                 Q249E   \n",
       "3   3     CBL                 N454D   \n",
       "4   4     CBL                 L399V   \n",
       "\n",
       "                                                Text  Gene_Share  \\\n",
       "0  Cyclin-dependent kinases (CDKs) regulate a var...           1   \n",
       "1   Abstract Background  Non-small cell lung canc...           1   \n",
       "2   Abstract Background  Non-small cell lung canc...           1   \n",
       "3  Recent evidence has demonstrated that acquired...           1   \n",
       "4  Oncogenic mutations in the monomeric Casitas B...           1   \n",
       "\n",
       "   Variation_Share Variation_0 Variation_1 Variation_2 Variation_3  \\\n",
       "0                1           T           r           u           n   \n",
       "1                1           W           8           0           2   \n",
       "2                1           Q           2           4           9   \n",
       "3                1           N           4           5           4   \n",
       "4                1           L           3           9           9   \n",
       "\n",
       "         ...        Variation_14 Variation_15 Variation_16 Variation_17  \\\n",
       "0        ...                   a            t            i            o   \n",
       "1        ...                                                              \n",
       "2        ...                                                              \n",
       "3        ...                                                              \n",
       "4        ...                                                              \n",
       "\n",
       "  Variation_18 Variation_19 Variation_20 Variation_21 Variation_22  \\\n",
       "0            n            s                                          \n",
       "1                                                                    \n",
       "2                                                                    \n",
       "3                                                                    \n",
       "4                                                                    \n",
       "\n",
       "  var_wild_plus_mut  \n",
       "0                Ts  \n",
       "1                W*  \n",
       "2                QE  \n",
       "3                ND  \n",
       "4                LV  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#commented for Kaggle Limits\n",
    "#Feature engineering 3\n",
    "#############################\n",
    "\n",
    "# for i in range(max(gene_len)):\n",
    "#     df_all['Gene_'+str(i)] = df_all['Gene'].map( \n",
    "#         lambda x: str(x[i]) if len(x)>i else '')\n",
    "for i in range(23):\n",
    "    df_all['Variation_'+str(i)] = df_all['Variation'].map(\n",
    "        lambda x: str(x[i]) if len(x)>i else '')\n",
    "    \n",
    "print(df_all.shape)\n",
    "\n",
    "##Add one more feature, a very effective one: the wildtype and mutant residues.\n",
    "df_all['var_wild_plus_mut'] = pd.Series(\n",
    "    map(lambda x: x[0]+x[-1], df_all['Variation']))\n",
    "df_all.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3578\n",
      "3578\n"
     ]
    }
   ],
   "source": [
    "#combine training genes and variants into a big list\n",
    "#used for text Search and Match later\n",
    "gen_var_lst = sorted(list(train.Gene.unique()) + list(train.Variation.unique()))\n",
    "print(len(gen_var_lst))\n",
    "\n",
    "#Using the first word for simplicity\n",
    "gen_var_lst = [ x.split(' ')[0] for x in gen_var_lst] \n",
    "print(len(gen_var_lst))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n",
      "700\n",
      "800\n",
      "900\n",
      "1000\n",
      "1100\n",
      "1200\n",
      "1300\n",
      "1400\n",
      "1500\n",
      "1600\n",
      "1700\n",
      "1800\n",
      "1900\n",
      "2000\n",
      "2100\n",
      "2200\n",
      "2300\n",
      "2400\n",
      "2500\n",
      "2600\n",
      "2700\n",
      "2800\n",
      "2900\n",
      "3000\n",
      "3100\n",
      "3200\n",
      "3300\n",
      "3400\n",
      "3500\n",
      "(4675, 3590)\n"
     ]
    }
   ],
   "source": [
    "#Feature engineering 3\n",
    "############################\n",
    "\n",
    "#count occurencies in the text\n",
    "i_ = 0\n",
    "for gen_var_lst_itm in gen_var_lst:\n",
    "    if i_ % 100 == 0: print(i_)\n",
    "    df_all['GV_'+str(gen_var_lst_itm)] = df_all['Text'].map(\n",
    "        lambda x: str(x).lower().count(str(gen_var_lst_itm).lower()))\n",
    "    i_ += 1\n",
    "    \n",
    "print(df_all.shape)\n",
    "#(4675, 3613)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4675, 3590) (3689,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(4675, 3590)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df_all.shape, y.shape)\n",
    "df_all_backup = df_all.copy()\n",
    "df_all_backup.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Gene', 'Variation', 'Text', 'Gene_Share', 'Variation_Share',\n",
       "       'Variation_0', 'Variation_1', 'Variation_2', 'Variation_3',\n",
       "       ...\n",
       "       'GV_Y98H', 'GV_Y98N', 'GV_YAP1', 'GV_YAP1-FAM118B', 'GV_YAP1-MAMLD1',\n",
       "       'GV_YAP1-TFE3', 'GV_YWHAE-ROS1', 'GV_ZC3H7B-BCOR', 'GV_ZNF198-FGFR1',\n",
       "       'GV_p61BRAF'],\n",
       "      dtype='object', length=3590)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3689, 3598) (986, 3598)\n",
      "Pipeline...\n",
      "(3689, 3684) (986, 3684)\n"
     ]
    }
   ],
   "source": [
    "from sklearn import preprocessing\n",
    "from sklearn import base\n",
    "from sklearn import pipeline\n",
    "from sklearn import feature_extraction\n",
    "from sklearn import decomposition\n",
    "from sklearn import model_selection\n",
    "from sklearn import metrics\n",
    "\n",
    "#Adding new features to df_all\n",
    "for c in df_all.columns:\n",
    "    if df_all[c].dtype == 'object':\n",
    "        if c in ['Gene','Variation']: \n",
    "            lbl = preprocessing.LabelEncoder()\n",
    "            df_all[c+'_lbl_enc'] = lbl.fit_transform(df_all[c].values)  \n",
    "            df_all[c+'_len']     = df_all[c].map(lambda x: len(str(x)))\n",
    "            df_all[c+'_words']   = df_all[c].map(lambda x: len(str(x).split(' ')))\n",
    "        elif c != 'Text':  #including 'var_wild_plus_mut'\n",
    "            lbl = preprocessing.LabelEncoder()\n",
    "            df_all[c] = lbl.fit_transform(df_all[c].values)\n",
    "        if c=='Text': \n",
    "            df_all[c+'_len']   = df_all[c].map(lambda x: len(str(x)))\n",
    "            df_all[c+'_words'] = df_all[c].map(lambda x: len(str(x).split(' '))) \n",
    "\n",
    "train = df_all.iloc[:len(train_var)] #iloc takes only integer indices\n",
    "test  = df_all.iloc[len(train_var):]\n",
    "print(train.shape, test.shape)\n",
    "\n",
    "class cust_regression_vals(base.BaseEstimator, base.TransformerMixin):\n",
    "    def fit(self, x, y=None):\n",
    "        return self\n",
    "    def transform(self, x):\n",
    "        x = x.drop(['Gene', 'Variation', 'ID','Text'],axis=1).values\n",
    "        return x\n",
    "\n",
    "class cust_txt_col(base.BaseEstimator, base.TransformerMixin):\n",
    "    def __init__(self, key):\n",
    "        self.key = key\n",
    "    def fit(self, x, y=None):\n",
    "        return self\n",
    "    def transform(self, x):\n",
    "        return x[self.key].apply(str)\n",
    "\n",
    "#Convert data to a representation suitable for training\n",
    "print('Pipeline...')\n",
    "fp = pipeline.Pipeline([(\n",
    "    'union', pipeline.FeatureUnion(\n",
    "        n_jobs = -1,\n",
    "        transformer_list = [\n",
    "            (\n",
    "                'standard', \n",
    "                cust_regression_vals()\n",
    "            ),\n",
    "            (\n",
    "                'pi1', \n",
    "                pipeline.Pipeline([\n",
    "                    ('Gene', cust_txt_col('Gene')),\n",
    "                    ('count_Gene', feature_extraction.text.CountVectorizer(analyzer=u'char', ngram_range=(1,8))),\n",
    "                    ('tsvd1', decomposition.TruncatedSVD(n_components=20, n_iter=25, random_state=12))\n",
    "                ])\n",
    "            ),\n",
    "            (\n",
    "                'pi2', pipeline.Pipeline([\n",
    "                    ('Variation', cust_txt_col('Variation')), \n",
    "                    ('count_Variation',feature_extraction.text.CountVectorizer(analyzer=u'char',ngram_range=(1,8))), \n",
    "                    ('tsvd2', decomposition.TruncatedSVD(n_components=20, n_iter=25, random_state=12))\n",
    "                ])\n",
    "            ),\n",
    "            #commented for Kaggle Limits\n",
    "            ('pi3', pipeline.Pipeline(\n",
    "                [\n",
    "                    ('Text', cust_txt_col('Text')), \n",
    "                    ('tfidf_Text', feature_extraction.text.TfidfVectorizer(ngram_range=(1, 2))), \n",
    "                    ('tsvd3', decomposition.TruncatedSVD(n_components=50, n_iter=25, random_state=12))])\n",
    "            )\n",
    "        ]\n",
    "    )\n",
    ")])\n",
    "\n",
    "#Put pipeline in use right here!\n",
    "train = fp.fit_transform(train)\n",
    "test  = fp.transform(test);      \n",
    "print(train.shape, test.shape)\n",
    "# (3652, 3621) (986, 3621)\n",
    "# (3652, 3707) (986, 3707)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4, 5, 6, 7, 8])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0]\ttrain-mlogloss:2.10012\tvalid-mlogloss:2.11597\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.459925\tvalid-mlogloss:0.930422\n",
      "[200]\ttrain-mlogloss:0.224214\tvalid-mlogloss:0.891717\n",
      "[300]\ttrain-mlogloss:0.119677\tvalid-mlogloss:0.903867\n",
      "Stopping. Best iteration:\n",
      "[212]\ttrain-mlogloss:0.207811\tvalid-mlogloss:0.891003\n",
      "\n",
      "10 0.89100330523\n",
      "[0]\ttrain-mlogloss:2.10077\tvalid-mlogloss:2.11079\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.466626\tvalid-mlogloss:0.867544\n",
      "[200]\ttrain-mlogloss:0.229284\tvalid-mlogloss:0.823113\n",
      "Stopping. Best iteration:\n",
      "[195]\ttrain-mlogloss:0.236766\tvalid-mlogloss:0.822128\n",
      "\n",
      "10 0.822127930029\n",
      "[0]\ttrain-mlogloss:2.10069\tvalid-mlogloss:2.11148\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.469715\tvalid-mlogloss:0.876609\n",
      "[200]\ttrain-mlogloss:0.228229\tvalid-mlogloss:0.827374\n",
      "[300]\ttrain-mlogloss:0.118664\tvalid-mlogloss:0.833457\n",
      "Stopping. Best iteration:\n",
      "[223]\ttrain-mlogloss:0.195779\tvalid-mlogloss:0.82542\n",
      "\n",
      "10 0.825420162159\n",
      "[0]\ttrain-mlogloss:2.10232\tvalid-mlogloss:2.10836\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.466238\tvalid-mlogloss:0.81755\n",
      "[200]\ttrain-mlogloss:0.231872\tvalid-mlogloss:0.767819\n",
      "[300]\ttrain-mlogloss:0.121488\tvalid-mlogloss:0.769103\n",
      "Stopping. Best iteration:\n",
      "[257]\ttrain-mlogloss:0.158198\tvalid-mlogloss:0.764296\n",
      "\n",
      "10 0.764295931418\n",
      "[0]\ttrain-mlogloss:2.09887\tvalid-mlogloss:2.11112\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.455632\tvalid-mlogloss:0.903909\n",
      "[200]\ttrain-mlogloss:0.228897\tvalid-mlogloss:0.865225\n",
      "[300]\ttrain-mlogloss:0.117323\tvalid-mlogloss:0.874009\n",
      "Stopping. Best iteration:\n",
      "[217]\ttrain-mlogloss:0.203738\tvalid-mlogloss:0.864171\n",
      "\n",
      "10 0.864170514088\n",
      "[0]\ttrain-mlogloss:2.09828\tvalid-mlogloss:2.11317\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.457238\tvalid-mlogloss:0.984116\n",
      "[200]\ttrain-mlogloss:0.21562\tvalid-mlogloss:0.954627\n",
      "Stopping. Best iteration:\n",
      "[186]\ttrain-mlogloss:0.23966\tvalid-mlogloss:0.951638\n",
      "\n",
      "10 0.951637984823\n",
      "[0]\ttrain-mlogloss:2.0998\tvalid-mlogloss:2.11152\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.476245\tvalid-mlogloss:0.911093\n",
      "[200]\ttrain-mlogloss:0.247494\tvalid-mlogloss:0.871616\n",
      "[300]\ttrain-mlogloss:0.1341\tvalid-mlogloss:0.876899\n",
      "Stopping. Best iteration:\n",
      "[228]\ttrain-mlogloss:0.208429\tvalid-mlogloss:0.869984\n",
      "\n",
      "10 0.869984310261\n",
      "[0]\ttrain-mlogloss:2.09992\tvalid-mlogloss:2.11131\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.454234\tvalid-mlogloss:0.900764\n",
      "[200]\ttrain-mlogloss:0.218093\tvalid-mlogloss:0.859181\n",
      "[300]\ttrain-mlogloss:0.112452\tvalid-mlogloss:0.866194\n",
      "Stopping. Best iteration:\n",
      "[230]\ttrain-mlogloss:0.178621\tvalid-mlogloss:0.856203\n",
      "\n",
      "10 0.85620257339\n",
      "[0]\ttrain-mlogloss:2.10116\tvalid-mlogloss:2.11063\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.46158\tvalid-mlogloss:0.885932\n",
      "[200]\ttrain-mlogloss:0.224658\tvalid-mlogloss:0.848456\n",
      "Stopping. Best iteration:\n",
      "[185]\ttrain-mlogloss:0.248272\tvalid-mlogloss:0.84751\n",
      "\n",
      "10 0.847510104719\n",
      "[0]\ttrain-mlogloss:2.09887\tvalid-mlogloss:2.11283\n",
      "Multiple eval metrics have been passed: 'valid-mlogloss' will be used for early stopping.\n",
      "\n",
      "Will train until valid-mlogloss hasn't improved in 100 rounds.\n",
      "[100]\ttrain-mlogloss:0.459232\tvalid-mlogloss:0.980764\n",
      "[200]\ttrain-mlogloss:0.229535\tvalid-mlogloss:0.964882\n",
      "Stopping. Best iteration:\n",
      "[153]\ttrain-mlogloss:0.314767\tvalid-mlogloss:0.959027\n",
      "\n",
      "10 0.959026619848\n",
      "CPU times: user 2h 49min 15s, sys: 8.29 s, total: 2h 49min 24s\n",
      "Wall time: 2h 49min 31s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "#######################\n",
    "####Run for production\n",
    "#######################\n",
    "\n",
    "#not for run,\n",
    "#keep for record\n",
    "names = ['ID','class1', 'class2', 'class3', 'class4', 'class5', 'class6', 'class7', 'class8', 'class9']\n",
    "denom = 0\n",
    "fold = 10 #Change to 5, 1 for Kaggle Limits\n",
    "for i in range(fold):\n",
    "    params = {\n",
    "        'eta': 0.05,\n",
    "        'max_depth': 6,\n",
    "        'min_child_weight':3,\n",
    "        'objective': 'multi:softprob', #multinomial logistics regression\n",
    "        'eval_metric': 'mlogloss',\n",
    "        'num_class': 9,\n",
    "        'seed': i,\n",
    "        'silent': True\n",
    "    }\n",
    "    x1, x2, y1, y2 = model_selection.train_test_split(train, y, test_size=0.18, random_state=i)\n",
    "    \n",
    "    watchlist = [(xgb.DMatrix(x1, y1), 'train'), (xgb.DMatrix(x2, y2), 'valid')]\n",
    "    \n",
    "    model = xgb.train(params, xgb.DMatrix(x1, y1), 1000,  watchlist, verbose_eval=100, early_stopping_rounds=100)\n",
    "    score1 = metrics.log_loss(y2, model.predict(xgb.DMatrix(x2), ntree_limit=model.best_ntree_limit), \n",
    "                              labels=list(range(9)))\n",
    "    print(i, score1)\n",
    "    #if score < 0.9:\n",
    "    if denom != 0:\n",
    "        pred   = model.predict(xgb.DMatrix(test), ntree_limit=model.best_ntree_limit+80)\n",
    "        preds += pred\n",
    "\n",
    "        train_pred   = model.predict(xgb.DMatrix(train), ntree_limit=model.best_ntree_limit+80)\n",
    "        train_preds += train_pred\n",
    "\n",
    "    else:\n",
    "        pred   = model.predict(xgb.DMatrix(test), ntree_limit=model.best_ntree_limit+80)\n",
    "        preds  = pred.copy()\n",
    "        \n",
    "        train_pred   = model.predict(xgb.DMatrix(train), ntree_limit=model.best_ntree_limit+80)        \n",
    "        train_preds  = train_pred.copy()\n",
    "        \n",
    "    denom += 1\n",
    "    submission = pd.DataFrame(pred, columns=['class'+str(c+1) for c in range(9)])\n",
    "    submission['ID'] = pid\n",
    "    #submission.to_csv('submission_stage2_10fold'  + str(i) + '.csv', index=False)\n",
    "preds       /= denom\n",
    "train_preds /= denom\n",
    "\n",
    "submission = pd.DataFrame(preds, columns=['class'+str(c+1) for c in range(9)])\n",
    "submission['ID'] = pid\n",
    "submission=submission[names]\n",
    "submission.to_csv('output/submission_stage2_all_no_gene_name_final.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
